### ISSUE TITLE
py4j locks jvm from exit

### ISSUE THREAD

**Issue by gnom7** [bartdag/py4j#302]
Hello, I try to use some python libraries in java app.
From java code I start python process with callback server and java gateway server in daemon thread in expectation that this will allow jvm to exit when my program finishes. But after my program finishes there is still one thread that locks jvm from exit, it is the thread from `ScheduledExecutorService` from `CallBackClient`, and although this scheduler performs just cleanup it is not running in daemon thread but closed explicitly by calling `GatewayServer#shutdown` or `CallbackClient#shutdown`, is this made intentionally? could we change it to be a daemon?

**Comment by bartdag** [comment]
That will unfortunately require a new parameter to not break backward compatibility, but this is clearly possible. Scheduled for 0.10.7.

**Comment by gnom7** [comment]
> If the GatewayServer is correctly shut down, it should not prevent the application from stopping: if you are experiencing this problem, I would like to get more details about your use case and what might be causing this issue.

That is true, but I'd like to free client code from calling server shutdown (it shouldn't be aware of library implementation and the fact it uses py4j internally).

> If you were to run GatewayServer in a daemon thread, the application could stop even if GatewayServer has not shutdown, which is not a standard behavior.

No, since `CallBackClient` has `private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);` and `start(false)` won't help (need to call rather `GatewayServer#shutdown` or `CallbackClient#shutdown` to kickoff `executor.shutdownNow();` code in `CallBackClient`).

> Finally, if you really want to run GatewayServer in a daemon thread, you can just call start(false) and run the GatewayServer in your own daemon thread. start(false) does not spawn a new thread.

That is what I have tried, and it really doesn't spawn new thread, but `CallBackClient` does (in it's cleanup scheduler).

It is easy to fix if we use daemon threads for that executor, it shouldn't break backward compatibility (I mean API won't change), since this is just a cleanup service with background logic, it runs only when app runs and it becomes unnecessary when app exits (so I cannot imagine it can break any existing code).

**Comment by daw3rd** [comment]
Hmm, seems more complicated than I was able to see, but if we just set the GatewayServer thread to be a daemon than all is well.   We are initiating the python startup from the Java side but want the jvm to exit immediately when main() returns.  Unfortunately, unless we set the thread to be a daemon, it waits for 30 seconds (or so).  The following addition at line 767 fixes the problem, but not sure this would be wanted all the time

![image](https://user-images.githubusercontent.com/5297044/70274235-c3b32280-1779-11ea-8510-54816869a20b.png)

We are using the ClientServer class so it would be nice if this needs to be configurable, it is exposed to ClientServer constructor somehow.

**Comment by daw3rd** [comment]
Per your request above, here is our use case.  We have an IClassifier interface that does not implement Closeable nor should it.  We have a python implementation of IClassifier that uses ClientServer to connect to python started in java.    The main() that creates the IClassifier doesn't know it has a PythonClassifier and so can't call its close() method.  In this scenario I don't know how to have  ClientServer .shutdown() called.  I tried a shutdownHook() but it isn't called until all non-daemon threads have exited (i.e. won't be called).  finalize() is not an option since something has a reference to the object that could do the shutdown.    Thanks for any recommendations or fixes.
